eav/queryset.py

```diff
@@ -1,47 +1,273 @@
-#!/usr/bin/env python
-#
-#    This software is derived from EAV-Django originally written and
-#    copyrighted by Andrey Mikhaylenko <http://pypi.python.org/pypi/eav-django>
-#
-#    This is free software: you can redistribute it and/or modify
-#    it under the terms of the GNU Lesser General Public License as published
-#    by the Free Software Foundation, either version 3 of the License, or
-#    (at your option) any later version.
-#
-#    This software is distributed in the hope that it will be useful,
-#    but WITHOUT ANY WARRANTY; without even the implied warranty of
-#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-#    GNU Lesser General Public License for more details.
-#
-#    You should have received a copy of the GNU Lesser General Public License
-#    along with EAV-Django.  If not, see <http://gnu.org/licenses/>.
+'''
+Queryset module.
 
+This module contains custom EavQuerySet class used for overriding
+relational operators and pure functions for rewriting Q-expressions.
+Q-expressions need to be rewritten for two reasons:
+
+    1. In order to hide implementation from the user and provide easy to use
+       syntax sugar, i.e.::
+
+           Supplier.objects.filter(eav__city__startswith='New')
+
+       instead of::
+
+           city_values = Value.objects.filter(value__text__startswith='New')
+           Supplier.objects.filter(eav_values__in=city_values)
+
+       For details see: ``eav_filter``.
+
+    2. To ensure that Q-expression tree is compiled to valid SQL.
+       For details see: ``rewrite_q_expr``.
+'''
+
+import sys
+from functools import reduce, wraps
+
+from django.db import models
+from django.db.models import Q
 from django.db.models.query import QuerySet
 
-from .decorators import eav_filter
+from .models import Attribute, Value
+
+
+def is_eav_and_leaf(expr, gr_name):
```

Comment:
@IwoHerka please either update docstring to include function parameters description, or consider using more verbose names. It took me a while reading through the rest of the module to understand `gr_name` means `generic_relation_name`.